### ISSUE TITLE
[Bug] Mixed Precision (float16) numerical instability in GroupNormalization with small epsilon

### ISSUE THREAD

**Issue by amadhan882** [keras-team/keras#22586]
## Description

While using keras.mixed_precision.set_global_policy("mixed_float16"), the GroupNormalization layer produces NaN outputs when the epsilon parameter is set to a value smaller than the precision limit of float16 (e.g., $1e-12$). Since the smallest positive normalized value in FP16 is approximately $6.1e-5$, a very small epsilon effectively acts as zero during variance normalization, leading to numerical collapse.

## Poc

```
import os
os.environ["KERAS_BACKEND"] = "tensorflow"
import keras
import numpy as np

keras.mixed_precision.set_global_policy("mixed_float16")
inputs = keras.Input(shape=(16, 16, 64))
# Epsilon 1e-12 causes NaN in FP16
x = keras.layers.GroupNormalization(groups=8, epsilon=1e-12)(inputs)
model = keras.Model(inputs, x)

extreme_data = np.random.uniform(low=60000, high=70000, size=(1, 16, 16, 64)).astype("float32")
output = model(extreme_data)
print(f"NaN detected: {np.any(np.isnan(output))}")


```

## Observed Result

```
NaN detected: True

```

## Expected Behavior

The layer should either:

* Automatically clip the epsilon to a safe minimum (e.g., 1e-7) when the policy is float16.

* Perform the normalization internal math in float32 regardless of the policy (Upcasting).

## Actual Behavior

The layer produces NaN values immediately when exposed to large input values or small epsilons in mixed_float16 mode.

## Notes
* This issue arises from float16 precision limits rather than user misuse.
* Other normalization layers (e.g., BatchNormalization) internally upcast to float32 to avoid similar issues.

**Comment by ChiragSW** [comment]
Found the Root cause:
With autocast=True (it is true in default), inputs were cast to float16 before reaching call(). Values exceeding float16's max (65504) overflowed to inf, causing NaN propagation through normalization math. The existing internal float32 upcast couldn't recover already-lost values.

**Comment by amadhan882** [comment]
Hi @ChiragSW,

Thanks again for the clarification — I followed up with an updated PoC to isolate the epsilon behavior.

```

import os
os.environ["KERAS_BACKEND"] = "tensorflow"

import keras
import numpy as np

# Enable mixed precision
keras.mixed_precision.set_global_policy("mixed_float16")

# Model with very small epsilon
inputs = keras.Input(shape=(16, 16, 64))
x = keras.layers.GroupNormalization(groups=8, epsilon=1e-12)(inputs)
model = keras.Model(inputs, x)

# Use SAFE float16 range (avoid overflow)
safe_data = np.random.uniform(
    low=1.0,
    high=10.0,
    size=(1, 16, 16, 64)
).astype("float32")

output = model(safe_data)

print("NaN detected:", np.any(np.isnan(output)))
print("Output dtype:", output.dtype)

```

I adjusted the input values to stay within the safe float16 range (1–10) to avoid overflow, while keeping epsilon very small (1e-12). Under these conditions:

* No NaNs were observed
* Output remained stable in float16

This confirms that the NaNs in the original report were due to float16 overflow rather than epsilon scaling.

Appreciate your insight — this helped clarify the root cause

**Comment by Whatsonyourmind** [comment]
The existing comment correctly identified the overflow issue, but it is worth noting that the epsilon problem and the overflow problem are two separate failure modes that both need fixing:

**Failure mode 1 (epsilon underflow)**: When epsilon < 6.1e-5 (the fp16 smallest normal), it rounds to zero. Then the normalization computes x / sqrt(var + 0) = x / sqrt(var), which is fine *unless* the variance is also very small, causing division by near-zero.

**Failure mode 2 (input overflow)**: Values > 65504 overflow to inf in fp16. Once inf enters the computation, mean and variance become inf/NaN regardless of epsilon.

The robust fix is to upcast the normalization internals to float32, matching what BatchNormalization already does:

```python
def call(self, inputs):
    input_dtype = inputs.dtype
    # Upcast for numerical stability
    x = ops.cast(inputs, 'float32')
    
    mean = ops.mean(x, axis=self.axis, keepdims=True)
    variance = ops.var(x, axis=self.axis, keepdims=True)
    x_norm = (x - mean) / ops.sqrt(variance + self.epsilon)
    
    # Apply gamma/beta in float32, then cast back
    output = self.gamma * x_norm + self.beta
    return ops.cast(output, input_dtype)
```

This is the standard approach in all major frameworks (PyTorch's GroupNorm upcasts to float32 internally). The epsilon should also be clamped to at least `np.finfo(np.float16).tiny` (~6e-5) when the compute dtype is float16, as a defense-in-depth measure.

**Comment by Lawson-Darrow** [comment]
The overflow mode here is still open after #22589 (which added the epsilon underflow warning). GroupNormalization autocasts inputs to float16 in `__call__` before the float32 upcast inside `call()`, so any value above 65504 is already inf by the time the normalization math runs. LayerNormalization sidesteps this by setting `self.autocast = False` in its constructor. Proposing the same one-line change for GroupNormalization so inputs keep their original dtype and the existing float32 upcast does its job. Happy to open a PR with a regression test if that sounds right.
